Gems/PrefabDependencyViewer/Code/Source/DirectedGraph.h

```diff
@@ -0,0 +1,231 @@
+        using TemplateId = AzToolsFramework::Prefab::TemplateId;
+
+        struct MetaData
+        {
+        public:
+            AZ_CLASS_ALLOCATOR(MetaData, AZ::SystemAllocator, 0);
+
+            MetaData() = default;
+
+            MetaData(TemplateId tid, const char* source)
+            {
+                SetTemplateId(tid);
+                SetSource(source);
+            }
+
+            TemplateId GetTemplateId()
+            {
+                return m_tid;
+            }
+
+            const char* GetSource()
+            {
+                return m_source;
+            }
+        private:
+            void SetTemplateId(const TemplateId tid)
+            {
+                m_tid = tid;
+            }
+            void SetSource(const char* source)
+            {
+                m_source = source;
+            }
+
+            TemplateId m_tid;
+            const char* m_source;
+        };
+
+        class Node
+        {
+        public:
+            AZ_CLASS_ALLOCATOR(Node, AZ::SystemAllocator, 0);
+
+            Node(TemplateId tid, const char* source, Node* parent = nullptr)
+            {
+                SetMetaData(tid, source);
+                SetParent(parent);
+            }
+
+            MetaData GetMetaData()
+            {
+                return m_metaData;
+            }
+
+            Node* GetParent()
+            {
+                return m_parent;
+            }
+
+            // In the future want to be able to edit the tree.
+            void SetParent(Node* parent)
+            {
+                m_parent = parent;
+            }
+
+        private:
+            void SetMetaData(TemplateId tid, const char* source)
+            {
+                m_metaData = MetaData(tid, source);
+            }
+
+            MetaData m_metaData;
+            Node* m_parent;
+        };
+
+        class DirectedGraph
+        {
+        public:
+            DirectedGraph()
+            {
+                m_nodes = NodeSet();
+                m_children = ChildrenMap();
+                m_root = nullptr;
+            }
+
+            void AddNode(Node* node)
+            {
+                m_nodes.insert(node);
+            }
+
+            void AddChild(Node* parent, Node* child)
+            {
+                if (parent)
+                {
+                    m_children[parent].insert(child);
+                    child->SetParent(parent);
+                }
+                else
+                {
+                    m_root = child;
+                }
+            }
+
+            Node* GetRoot() const
+            {
+                return m_root;
+            }
+
+            NodeSet GetChildren(Node* parent) const
+            {
+                AZ_Assert(parent, "Nullptr can't have children");
+
+                auto iter = m_children.find(parent);
+                if (iter != m_children.end())
+                {
+                    return iter->second;
+                }
+
+                return NodeSet();
```

Comment:
You can avoid repeating the return type by using a braced initializer instead
```suggestion
                return {};
```

Reply:
Done